pytorch-qa/app.py

```diff
@@ -0,0 +1,412 @@
+#         # return response
+#         return next(streamer)
+
+#     @property
+#     def _identifying_params(self) -> Mapping[str, Any]:
+#         return {"name_of_model": self.model_name}
+
+#     @property
+#     def _llm_type(self) -> str:
+#         return "custom"
+
+# llm_predictor = LLMPredictor(llm=CustomLLM())
+#############################################################################
+
+hf_embed_model = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-xl")
+embed_model = LangchainEmbedding(hf_embed_model)
+service_context = ServiceContext.from_defaults(
+    llm_predictor=llm_predictor, prompt_helper=prompt_helper, embed_model=embed_model
+)
+
+
+def load_index():
+
+    # load data
+    storage_context = StorageContext.from_defaults(
+        docstore=SimpleDocumentStore.from_persist_dir(persist_dir="./pytorch_docs_512"),
+        vector_store=SimpleVectorStore.from_persist_dir(persist_dir="./pytorch_docs_512"),
+        index_store=SimpleIndexStore.from_persist_dir(persist_dir="./pytorch_docs_512"),
+    )
+    index = load_index_from_storage(storage_context, service_context=service_context)
+
+    return index
+
+
+def load_tools(index):
+
+    # Insert your huggingface token here
+    API_TOKEN = ""
+
+    class get_query(BaseModel):
+        query: str = Field(
+            default=None,
+            description="The name of the model and task for which you want the datasets.",
+        )
+
+    @tool
+    def get_datasets(query: str) -> str:
+        """Fetches dataset for a given model and task"""
+        print("Log::::::::::::::::::::::::::::", query, type(query))
+        pattern = r"model=(.*?)\s*(?:and\s*)?task=(.*?)$"
+        match = re.search(pattern, query)
+        model_name = match.group(1)
+        task = match.group(2)
+        print("Log::::::::::::::::::::::::::::", model_name, task)
+
+        filter = ModelFilter(library="pytorch", model_name=model_name, task=task)
+
+        list_of_models = list_models(filter=filter, token=API_TOKEN)
+        if not list_of_models:
+            return f"{model_name} Not Found!"
+        filtered_model = [model for model in list_of_models if model.id == model_name]
+        datasets = [
+            string.split(":")[1] for string in filtered_model[0].tags if "dataset:" in string
+        ]
+        if not datasets:
+            return f"No Datasets available for {model_name}"
+        print("Log::::::::::::::::::::::::::::", ",".join(datasets))
+        return ",".join(datasets)
+
+    search_tools = [
+        StructuredTool.from_function(
+            func=get_datasets,
+            name="Get Model Datasets",
+            args_schema=get_query,
+            return_direct=True,
+            description="A search engine. Useful for when you need to answer questions about fetching datasets. For example the input should be like model=model and task=task. Do not use this tool with the same input/query",
```

Comment:
In the description replace "A search engine" with "Tool for searching huggingface datasets"